envoy/network/listener.h

```diff
@@ -238,9 +240,9 @@ class ListenerConfig {
   virtual const std::string& name() const PURE;
 
   /**
-   * @return ListenerInfo& description of the listener.
+   * @return ListenerInfoConstSharedPtr& description of the listener.
    */
-  virtual const ListenerInfo& listenerInfo() const PURE;
+  virtual const ListenerInfoConstSharedPtr& listenerInfo() const PURE;
```

Comment:
Is this change necessary?

Reply:
Yeah, I want to use shared ptr since it's difficult to ensure references are safe and listener info is already stored as a shared ptr.

Reply:
Sorry, I actually didn't get it. The listener info is owned by the factory context and will always have longer lifetime than the `ListenerConfig`. And I will shared the ownership with the connection info provider which ensure I can always access it by the valid connection info provider. 

So, I didn't see the necessity to change this method, except you want to get the ownership of listener info by this method (but I didn't see related usage in this PR).

Reply:
I'm not confident `ActiveTcpListener` stream doesn't outlive `ListenerConfig`. The relationship stored is via direct pointers, and it's guaranteed to exist at the connection acceptance, but not for the whole duration of a stream, no?

Reply:
AFAIK, the drain manager will ensure all connection be closed before actually destroy a filter chain in the main thread. I am not very sure because long time has gone after last time I read related code. 

But it doesn't matter. Note the connection info will own the listener info also now and the connection info will be owned by the `StreamInfo` which have same lifetime with the related stream.

Reply:
Thanks, you are right, we can simplify this PR a lot if we assume lifetime works.